tokio/tests/sync_rwlock.rs

```diff
@@ -0,0 +1,135 @@
+#![warn(rust_2018_idioms)]
+
+use std::sync::Arc;
+
+use futures_util::future::FutureExt;
+use futures_util::stream;
+use futures_util::stream::StreamExt;
+
+use tokio::sync::{Barrier, RwLock};
+use tokio_test::task::spawn;
+use tokio_test::{assert_pending, assert_ready};
+
+// multiple reads should be Ready
+#[test]
+fn read_shared() {
+    let rwlock = RwLock::new(100);
+
+    let mut t1 = spawn(rwlock.read());
+    assert_ready!(t1.poll());
+    let mut t2 = spawn(rwlock.read());
+    assert_ready!(t2.poll());
+}
+
+// When there is an active shared owner, exclusive access should not be possible
+#[test]
+fn write_shared_pending() {
```

Comment:
How about an additional test like this one but in the opposite order?  It gets a write lock and then, without dropping the guard, attempts to get a read lock but fails.  A third test that gets a write lock and then tries to write again would also be in order.